Added a REST API with authentication for retrieving balances and goals

Laugslander opened the pull request:
Implemented a REST API using the JSON API standard for retrieving Kudo balances and goals. The API is protected with simple token based authentication.

pascalw commented on app/controllers/api/v1/authentication_controller.rb:
```diff
@@ -0,0 +1,32 @@
+  respond_to :json
+
+  def retrieve_api_token
+    if validate_jwt_token(params[:jwt_token])
```
Who's going to call this API? It requires a JWT token - who is providing this token?

Laugslander replied:
The Kudo-o-Mobile app will call this function. The app provides a JWT token that it receives from signing in using Google OAuth2. The backend validates this token.

pascalw commented on app/models/user.rb:
```diff
@@ -1,33 +1,37 @@
+  end
+
+  def self.from_api_token_request(params)
+    user = User.where(uid: params['uid']).first_or_create(params.except(:jwt_token))
```
Does it make sense to create a user during an API token request? Sounds fishy 🤔

Laugslander replied:
Te API token is stored in the User table in database. When a new user signs into the the mobile app (before signing into the web app), he doesn't have a User record yet.

pascalw commented on spec/requests/api/v1/balances_controller_spec.rb:
```diff
@@ -0,0 +1,215 @@
+      it 'returns all balances' do
+        get request, headers: {'Api-Token': user.api_token}
+
+        expect(json['data'][0]['id']).to eq(balance1.id.to_s)
```
This is OK, but since this API is going to be a contract (between the Kudo server and some client(s)) I would rather have the expectations more exact, I.E. matching the entire response. This to prevent accidental API breakage in the future, for example by removing a field not included in the expectation currently.

pascalw commented on spec/requests/api/v1/balances_controller_spec.rb:
```diff
@@ -0,0 +1,215 @@
+          post request,
+               headers: {'Api-Token': user.api_token, 'Content-Type': 'application/vnd.api+json'},
+               params: {data: {type: 'balances', attributes: {name: balance.name, current: balance.current}}}.to_json
+        }.to change {Balance.count}
```
Also here I think you could be a bit more precise in your expectation, to verify that all JSON fields are mapped properly. (So include all fields in the expectation).

pascalw commented on spec/requests/api/v1/balances_controller_spec.rb:
```diff
@@ -0,0 +1,215 @@
+              params: {data: {type: 'balances', id: balance.id, attributes: {name: 'edited name', current: true}}}.to_json
+
+        expect(json['data']['attributes']['name']).to eq('edited name')
+        expect(json['data']['attributes']['current']).to eq(true)
```
Also here it would be good to test that the PATCH request updates only the fields that are passed in the request body and leaves any field _not_ included untouched.

pascalw commented on spec/support/request_helpers.rb:
```diff
@@ -0,0 +1,11 @@
+      JSON.parse(response.body).with_indifferent_access
+    end
+
+    def expect_unauthorized
```
Since the API is a contract, please include the HTTP status code in your expectation as well. This is very important for REST clients.

pascalw commented on app/controllers/api/v1/goals_controller.rb:
```diff
@@ -0,0 +1,9 @@
+class Api::V1::GoalsController < Api::V1::ApiController
+  def next_amount
```
Is it useful to have the next amount and name separate? My guess would be that API clients would need both the name and amount at the same time, so you might want to combine this into one API call.

Also the path of this resources perhaps could be `GET /api/v1/goals/next`. Where `next` is a symbolic primary key that identifies the next goal.

pascalw commented on spec/requests/api/v1/goals_controller_spec.rb:
```diff
@@ -0,0 +1,250 @@
+    let! (:goal1) {create(:goal)}
+    let! (:goal2) {create(:goal)}
+
+    context 'with valid api-token' do
```
What would happen when there is no next goal? Might want to add a test for this case too.

pascalw commented on Gemfile:
```diff
@@ -54,6 +54,11 @@ gem 'redcarpet', '~> 3.3', '>= 3.3.4'
 
 gem 'timecop', '~> 0.8.1'
 
+gem 'jsonapi-resources'
```
It's generally preferable to specify versions of your dependencies explicitly. If you don't, then gems will likely get unintentionally updated at some point, potentially breaking your app.

It's fine to leave a little fuzzy about versions, for example by specifying `'gem 'jsonapi-resources', '~> 0.9'. This would allow 'safe' updates in the `0.9.x` range, but disallow larger updates like `0.10` or `1.0`.

Laugslander commented:
- Improved and refactored the naming of the balance- and goal specific non-CRUD API controller actions. 

- Improved and refactored the integration tests by making the response tests more specific and adding tests for HTTP status codes.

- Added versions to the API related gems in the Gemfile file (also created an issue for this).
